ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/OptimisticHeadValidator.java

```diff
@@ -79,6 +82,19 @@ public OptimisticHeadValidator(
     return SafeFuture.COMPLETE;
   }
 
+  private void execute() {
+    verifyOptimisticHeads();
+    if (reexecutingExecutionPayload.compareAndSet(false, true)) {
+      LOG.debug("re-executing execution payloads for queued blocks");
+      SafeFuture.asyncDoWhile(this::reexecuteQueuedExecutionPayloadRetry)
+          .always(
+              () -> {
+                LOG.debug("re-executing completed");
+                reexecutingExecutionPayload.set(false);
+              });
+    }
```

Comment:
I think these are two separate functions and should live in separate classes.  So I'd leave this `OptimisticHeadValidator` as is and have either `BlockManager` or probably a new class it delegates to try to reprocess blocks that had to be queued because they couldn't be optimistically imported.